### PR TITLE
Add product photo detector test

### PR DESCRIPTION
**Pull request by ralucatudor**
Using the FakeProductDetectionAPIImpl, which implements ProductDetectionAPI (see [previous PR](https://github.com/googleinterns/step256-2020/pull/32)), test ProductPhotoDetector's buildShoppingQuery.

### CODE REVIEW

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+import org.junit.runners.JUnit4;
+
+/** 
+ * Tests the behaviour of ProductPhotoDetector, namely the shopping query returned and the 
+ * exceptions thrown.
+ */
+@RunWith(JUnit4.class)
+public final class ProductPhotoDetectorTest {
+
+  private static final byte[] IMAGE_BYTES = "Fake image".getBytes();
```
Comment (ChangchengXiao) [inline src/test/java/com/google/sps/ProductPhotoDetectorTest.java:36]: Do you need to do this? Could a new byte [1] be sufficient?

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  private ProductPhotoDetector productPhotoDetector;
+  private FakeProductDetectionAPI fakeProductDetectionAPI;
```
Comment (ChangchengXiao) [inline src/test/java/com/google/sps/ProductPhotoDetectorTest.java:38]: I think a better name could be "FakeProductDetectionAPIImpl"

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  public void setUp() {
+    productPhotoDetector = new ProductPhotoDetector();
+    fakeProductDetectionAPI = new FakeProductDetectionAPI();
+  }
+
+  @Test
+  public void completeDetectionData() throws Exception {
+    List<String> labels = new ArrayList();
+    labels.add("Shoe");
+    labels.add("Sneaker");
```
Comment (ChangchengXiao) [inline src/test/java/com/google/sps/ProductPhotoDetectorTest.java:50]: Using ImmutableList is better. e.g.
https://www.geeksforgeeks.org/immutable-list-in-java/#:~:text=ImmutableList%2C%20as%20suggested%20by%20the,the%20List%2C%20UnsupportedOperationException%20is%20thrown.

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    List<String> logos = new ArrayList();
+    logos.add("Nike");
+    List<String> colors = new ArrayList();
+    colors.add("Black");
+
+    ProductDetectionData productDetectionData = new ProductDetectionData(labels, logos, colors);
+    fakeProductDetectionAPI.setReturnValue(productDetectionData);
```
Comment (ChangchengXiao) [inline src/test/java/com/google/sps/ProductPhotoDetectorTest.java:57]: It will be nice if you could create a private method to do such setups,  e.g.

"setProductDetectionResult(ImmutableList<String> labels, ....)

By doing this:
#1 your unit test will be easier to read
#2 you can avoid boilerplate code among different tests.

**Review by ChangchengXiao — COMMENTED** [review]
Overall looks good!

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java (hunk 4 above)
Comment (ralucatudor) [reply src/test/java/com/google/sps/ProductPhotoDetectorTest.java]: Indeed, solved!

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java (hunk 3 above)
Comment (ralucatudor) [reply src/test/java/com/google/sps/ProductPhotoDetectorTest.java]: I made the list immutable, done! But I followed the example for JDK 8 from [here](https://docs.oracle.com/javase/9/core/creating-immutable-lists-sets-and-maps.htm#JSCOR-GUID-202D195E-6E18-41F6-90C0-7423B2C9B381) and used `Collections.unmodifiableList(...)`. And for the empty lists, I used `Collections.emptyList()`, so these are immutable as well

Comment (ralucatudor) [reply src/test/java/com/google/sps/ProductPhotoDetectorTest.java]: @ChangchengXiao is this approach okay?

**Review by saakshisaxena — APPROVED** [review]
LGTM 👍

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java (hunk 3 above)
Comment (ralucatudor) [reply src/test/java/com/google/sps/ProductPhotoDetectorTest.java]: Update: I now use Guava's ImmutableList :)

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  @Test
+  public void emptyDetectionData() throws Exception {
+    ImmutableList<String> labels = ImmutableList.of();
+    ImmutableList<String> logos = ImmutableList.of();
+    ImmutableList<String> colors = ImmutableList.of();
+
+    ProductPhotoDetector productPhotoDetector = initProductPhotoDetector(labels, logos, colors);
+
+    Assertions.assertThrows(PhotoDetectionException.class, () -> {
+        productPhotoDetector.buildShoppingQuery(IMAGE_BYTES);
```
Comment (saakshisaxena) [inline src/test/java/com/google/sps/ProductPhotoDetectorTest.java:77]: I think it'll be better if we can check the  message of the exception, just to be sure we are throwing the exception from the right place.
I looked at the example from [this link](https://www.baeldung.com/junit-assert-exception).

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** 
+ * Tests the behaviour of {@link ProductPhotoDetector#buildShoppingQuery(byte[])}.
+ */
+@RunWith(JUnit4.class)
+public final class ProductPhotoDetectorTest {
+
+  private static final byte[] IMAGE_BYTES = new byte[1];
```
Comment (saakshisaxena) [inline src/test/java/com/google/sps/ProductPhotoDetectorTest.java:34]: I think we can also write a test case for when IMAGE_BYTES are null :) 
```suggestion
  private static final byte[] NULL_IMAGE_BYTES = new byte[0];
  private static final byte[] IMAGE_BYTES = new byte[1];
```

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    fakeProductDetection = new FakeProductDetectionAPIImpl();
+  }
+
+  /** Sets product detection results and initializes ProductPhotoDetector instance. */
+  private ProductPhotoDetector initProductPhotoDetector(ImmutableList<String> labels, 
+      ImmutableList<String> logos, ImmutableList<String> colors) {
+    ProductDetectionData productDetectionData = ProductDetectionData.create(labels, logos, colors);
+    fakeProductDetection.setReturnValue(productDetectionData);
+
+    ProductPhotoDetector productPhotoDetector = new ProductPhotoDetector(fakeProductDetection);
```
Comment (saakshisaxena) [inline src/test/java/com/google/sps/ProductPhotoDetectorTest.java:49]: What do you think about making `ProductPhotoDetector productPhotoDetector ` a class variable (cause all of your methods are using a `ProductPhotoDetector` instance that are getting initialised separately anyways) ?

**Review by saakshisaxena — APPROVED** [review]
On a second note, now that I started writing my tests :P , I would like to suggest a couple of things 😄

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java (hunk 7 above)
Comment (ralucatudor) [reply src/test/java/com/google/sps/ProductPhotoDetectorTest.java:49]: Yes, I thought about it! At first, it was a class field and it was initialized in the setUp(). But, after adding the API class in the constructor, I decided to not have it as a field, as I have to create a new instance of the ProductPhotoDetector everytime, after I set the behaviour of of the FakeProductDetectionAPIImpl

File: src/test/java/com/google/sps/ProductPhotoDetectorTest.java (hunk 5 above)
Comment (ralucatudor) [reply src/test/java/com/google/sps/ProductPhotoDetectorTest.java:77]: Yes, I actually believe I tried this and didn't work because I was using JUnit 4 plus the Jupiter. I'll see if I can make it work now

**Review by ChangchengXiao — APPROVED** [review]
